mono/utils/mono-threads.c

```diff
@@ -1835,3 +1911,181 @@ mono_thread_info_get_tools_data (void)
+		MONO_ENTER_GC_SAFE;
+		mono_os_sem_timedwait (&thread->work_queue_sem, 1000, MONO_SEM_FLAGS_NONE);
+		MONO_EXIT_GC_SAFE;
+		utility_thread_handle_inbox (thread);
+	}
+
+	/* Drain any remaining entries on shutdown. */
+	while (utility_thread_handle_inbox (thread));
+
+	mono_os_sem_destroy (&thread->work_queue_sem);
+
+	thread->callbacks.cleanup (thread->state_ptr);
+
+	return NULL;
+}
+
+MonoUtilityThread *
+mono_utility_thread_launch (size_t payload_size, MonoUtilityThreadCallbacks *callbacks)
+{
+	MonoUtilityThread *thread = g_malloc0 (sizeof (MonoUtilityThread));
+	size_t entry_size = offsetof (UtilityThreadQueueEntry, payload) + payload_size;
+
+	thread->message_block_size = mono_pagesize ();
+	thread->payload_size = payload_size;
+	thread->callbacks = *callbacks;
+
+	mono_lock_free_queue_init (&thread->work_queue);
+	mono_lock_free_allocator_init_size_class (&thread->message_size_class, entry_size, thread->message_block_size);
+	mono_lock_free_allocator_init_allocator (&thread->message_allocator, &thread->message_size_class, MONO_MEM_ACCOUNT_OTHER);
+	mono_os_sem_init (&thread->work_queue_sem, 0);
+	mono_atomic_store_i32 (&thread->run_thread, 1);
+
+	if (!mono_native_thread_create (&thread->thread_id, utility_thread, thread))
+		g_error ("Could not create utility thread");
+
+	return thread;
+}
+
+static void
+mono_utility_thread_send_internal (MonoUtilityThread *thread, UtilityThreadQueueEntry *entry)
+{
+	mono_lock_free_queue_node_init (&entry->node, FALSE);
+	mono_lock_free_queue_enqueue (&thread->work_queue, &entry->node);
+	mono_os_sem_post (&thread->work_queue_sem);
+}
+
+void
+mono_utility_thread_send (MonoUtilityThread *thread, gpointer message)
+{
+	int small_id = mono_thread_info_get_small_id ();
+	if (small_id < 0) {
+#if 1
+		fprintf (stderr, "Dropping message send because thread not attached yet\n");
+#endif
+		return;
+	} else if (!thread->run_thread) {
+#if 1
+		fprintf (stderr, "Dropping message send because thread killed\n");
+#endif
+		return;
+	}
+
+	UtilityThreadQueueEntry *entry = mono_lock_free_alloc (&thread->message_allocator);
+	entry->response_sem = NULL;
+	entry->thread = thread;
+	memcpy (entry->payload, message, thread->payload_size);
+	mono_utility_thread_send_internal (thread, entry);
+}
+
+gboolean
+mono_utility_thread_send_sync (MonoUtilityThread *thread, gpointer message)
+{
+	int small_id = mono_thread_info_get_small_id ();
+	if (small_id < 0) {
+#if 1
+		fprintf (stderr, "Dropping message send because thread not attached yet\n");
+#endif
+		return FALSE;
+	} else if (!thread->run_thread) {
+#if 1
+		fprintf (stderr, "Dropping message send because thread killed\n");
+#endif
+		return FALSE;
+	}
+
+	MonoSemType sem;
+	mono_os_sem_init (&sem, 0);
+
+	UtilityThreadQueueEntry *entry = mono_lock_free_alloc (&thread->message_allocator);
+	gboolean done;
+
+	entry->finished = &done;
+	entry->response_sem = &sem;
+	entry->thread = thread;
+	memcpy (entry->payload, message, thread->payload_size);
```

Comment:
I'm a little bit confused about the memory management of this utility thread mechanism.

So you send it messages - it will allocate its own copy and then queue that for the utility thread to process.  When the queue processes an entry, it will pass a pointer to the copy that it owns to the `command` callback but then it doesn't free the entry.  And the `handle_command` that you wrote for the flight recorder doesn't free the entry that's passed to it either.

So who is freeing the copy of the message when it's pulled off the queue?

Reply:
The message is passed to the hazard pointer freeing mechanism. The payload blob is just a static memory blob that we memcpy into. There's no dynamic memory allocation outside of the queue, after we allocate our ring buffer's memory array.

Reply:
utility_thread_handle_inbox pops the message off of the queue, passes the pointer to the static memory to the flight recorder so it can memcpy out of it, and then frees the message.

Reply:
I was worried about increasing all the data motion, but in a cachegrind/ubuntu benchmark it was much faster than even a single cache miss from pointers so data traffic was kind of a moot point.

Reply:
Ah ok. I missed that the entry is variably-sized.

So if the message owned some allocated memory, the call to send would transfer ownership to the queue, and the `command` callback would be responsible for deallocating.

Reply:
Yeah, that memory lifetime contract would be between the sender and the command callback.